File: python/xiangqi-bot/src/xiangqi_bot/game/moves.py
"""走法推断、应用与格式化（纯函数）。

输入棋盘/变动数据，输出走法或新状态；不做 IO、不访问 self。
"""

from __future__ import annotations

from xiangqi_bot.board import PIECE_CN, Board, grid_to_square, piece_color, piece_label
from xiangqi_bot.game.state import Change, Move, Side


def infer(changes: list[Change]) -> Move | None:
    """从 2 格变动推断一步棋：一起一落、棋子相同。"""
    if len(changes) != 2:
        return None
    left: tuple[int, int, str] | None = None
    arrived: tuple[int, int, str | None, str] | None = None
    for r, c, old, new in changes:
        if old is not None and new is None:
            left = (r, c, old)
        elif new is not None:
            arrived = (r, c, old, new)
    if left is not None and arrived is not None and arrived[3] == left[2]:
        return Move((left[0], left[1]), (arrived[0], arrived[1]), left[2], arrived[2])
    return None


def apply(board: Board, move: Move, clock: int) -> int:
    """把走法写入 board（起子格清空、落子格写棋子），返回新的 halfmove_clock。

    吃子归零，非吃 +1。我方/敌方走棋共用同一个函数。
    """
    (r1, c1), (r2, c2) = move.src, move.dst
    board[r1][c1] = None
    board[r2][c2] = move.piece
    return 0 if move.captured is not None else clock + 1


def matches(move: Move, expected: Move) -> bool:
    """走法是否与引擎着法完全吻合（起点/终点/棋子）。"""
    return move.src == expected.src and move.dst == expected.dst and move.piece == expected.piece


def format_move(move: Move, my_side: Side) -> str:
    """格式化走棋日志（红/黑方 + 棋子 + 记谱 + 吃子）。"""
    piece, captured = move.piece, move.captured
    color_cn = "红" if piece_color(piece) == Side.RED else "黑"
    from_sq = grid_to_square(*move.src, my_side)
    to_sq = grid_to_square(*move.dst, my_side)
    capture_note = f"（吃{piece_label(captured)}）" if captured else ""
    return f"{color_cn}方走{PIECE_CN[piece]}：{from_sq} -> {to_sq}{capture_note}"


def format_changes(changes: list[Change], my_side: Side) -> list[str]:
    """格式化逐格变动日志（含标题行）。"""
    lines = [f"棋子变动（{len(changes)} 格）"]
    for r, c, old, new in changes:
        old_name = piece_label(old) if old else "空"
        new_name = piece_label(new) if new else "空"
        lines.append(f"变化：{grid_to_square(r, c, my_side)} {old_name} -> {new_name}")
    return lines

File: android/chess_bot/app/src/main/java/com/chess/bot/game/moves.kt
package com.chess.bot.game

/** 走法推断、应用与格式化（纯函数，移植 python moves.py）。 */

/** 从 2 格变动推断一步棋：一起一落、棋子相同；不吻合返回 null。 */
fun inferMove(changes: List<Change>): Move? {
    if (changes.size != 2) return null
    var left: Triple<Int, Int, String>? = null
    var arrived: Quadr? = null
    for ((r, c, old, new) in changes) {
        if (old != null && new == null) left = Triple(r, c, old)
        else if (new != null) arrived = Quadr(r, c, old, new)
    }
    val l = left ?: return null
    val a = arrived ?: return null
    if (a.new != l.third) return null
    return Move((l.first to l.second), (a.r to a.c), l.third, a.old)
}

private data class Quadr(val r: Int, val c: Int, val old: String?, val new: String)

/** 把走法写入 board（起子清空、落子写子），返回新的 halfmoveClock：吃子归零，非吃 +1。 */
fun applyMove(board: Board, move: Move, clock: Int): Int {
    val (r1, c1) = move.src
    val (r2, c2) = move.dst
    board[r1][c1] = null
    board[r2][c2] = move.piece
    return if (move.captured != null) 0 else clock + 1
}

/** 走法是否与引擎着法完全吻合（起点/终点/棋子）。 */
fun moveMatches(move: Move, expected: Move): Boolean =
    move.src == expected.src && move.dst == expected.dst && move.piece == expected.piece

/** 格式化走棋日志：红方走帥：h2 -> e2（吃砲）。 */
fun formatMove(move: Move, mySide: Side): String {
    val colorCn = if (pieceColor(move.piece) == Side.RED) "红" else "黑"
    val from = gridToSquare(move.src.first, move.src.second, mySide)
    val to = gridToSquare(move.dst.first, move.dst.second, mySide)
    val captureNote = move.captured?.let { "（吃${pieceLabel(it)}）" } ?: ""
    return "${colorCn}方走${PIECE_CN[move.piece]}：$from -> $to$captureNote"
}

/** 格式化逐格变动日志（含标题行）。 */
fun formatChanges(changes: List<Change>, mySide: Side): List<String> {
    val lines = mutableListOf("棋子变动（${changes.size} 格）")
    for ((r, c, old, new) in changes) {
        val sq = gridToSquare(r, c, mySide)
        lines.add("变化：$sq ${old?.let { pieceLabel(it) } ?: "空"} -> ${new?.let { pieceLabel(it) } ?: "空"}")
    }
    return lines
}
